PR: https://github.com/odoo/odoo/pull/229706

From: 2dff65ab8b5a9db21d5b476065a72755cc4625be
From: romc-odoo
Date: 2026-01-16 11:09:08

Breaking data model changes scores: del:15.4 + add:0, change matches:
-    attendance_ids_1st_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 1st Week',
-    attendance_ids_2nd_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 2nd Week',
-    schedule_type = fields.Selection(
-    duration_based = fields.Boolean("Attendance based on duration", help="The hours will be centered around 12:00 to cover the duration for the day")
-    flexible_hours = fields.Boolean(string="Flexible Hours",
-    two_weeks_calendar = fields.Boolean(string="Calendar in 2 weeks mode")
-    two_weeks_explanation = fields.Char('Explanation', compute="_compute_two_weeks_explanation")
-    tz = fields.Selection(
-    tz_offset = fields.Char(compute='_compute_tz_offset', string='Timezone offset')
-    name = fields.Char(required=True)
-    duration_days = fields.Float(compute='_compute_duration_days', string='Duration (days)', store=True, readonly=False)
-    duration_based = fields.Boolean(related="calendar_id.duration_based")
+    duration_based = fields.Boolean(compute='_compute_duration_based', store=True)
-    week_type = fields.Selection([
-    two_weeks_calendar = fields.Boolean("Calendar in 2 weeks mode", related='calendar_id.two_weeks_calendar')

Total Changes: 857

[IMP] resource: refactor working schedule

This commit includes a refactoring of the working schedule (resource calendar).
The following changes have been made:
* Display required full time + work time rate in debug mode
* Always define the amount of hours per day for resource.calendar.attendance
* Remove the 2 weeks calendar feature, that will be refactored in another PR
* Remove the name, duration days and day period fields on resource.calendar.attendance.
* Remove the timezone from the calendar. Only the one on the resource/employee/version is used
* Flexible calendars don't exist anymore. Instead, the resource/employee doesn't have a calendar but specify hours per week/day
* No more "lunch". It means that an employee that, for example, doesn't check out during lunch time will be counted as working during these hours.

task-5124688

closes odoo/odoo#229706

Related: odoo/enterprise#96139
Related: odoo/upgrade#9030
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/resource/models/res_company.py
+++ b/addons/resource/models/res_company.py
@@ -1,7 +1,7 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import api, fields, models, _
+from odoo.addons.base.models.res_partner import _tz_get
 
 
 class ResCompany(models.Model):
@@ -11,6 +11,9 @@ class ResCompany(models.Model):
         'resource.calendar', 'company_id', 'Working Hours')
     resource_calendar_id = fields.Many2one(
         'resource.calendar', 'Default Working Hours', ondelete='restrict')
+    tz = fields.Selection(
+        _tz_get, string='Timezone', required=True,
+        default=lambda self: self.env.context.get('tz') or self.env.user.tz or 'UTC')
 
     @api.model
     def _init_data_resource_calendar(self):

--- a/addons/resource/models/res_users.py
+++ b/addons/resource/models/res_users.py
@@ -16,12 +16,9 @@ class ResUsers(models.Model):
     def write(self, vals):
         rslt = super().write(vals)
 
-        # If the timezone of the admin user gets set on their first login, also update the timezone of the default working calendar
+        # If the timezone of the admin user gets set on their first login, also update the timezone of the company
         if (vals.get('tz') and len(self) == 1 and not self.env.user.login_date
             and self.env.user == self.env.ref('base.user_admin', False) and self == self.env.user):
-            if self.resource_calendar_id:
-                self.resource_calendar_id.tz = vals['tz']
-            else:
-                self.env.ref('resource.resource_calendar_std', False).tz = vals['tz']
-
+            if self.company_id:
+                self.company_id.tz = vals['tz']
         return rslt

--- a/addons/resource/models/resource_calendar.py
+++ b/addons/resource/models/resource_calendar.py
@@ -2,32 +2,21 @@
 
 from collections import defaultdict
 from datetime import datetime, time, timedelta, UTC
-from functools import partial
 from itertools import chain
-from typing import NamedTuple
+from functools import partial
 from zoneinfo import ZoneInfo
 
 from dateutil.relativedelta import relativedelta
 from dateutil.rrule import DAILY, rrule
 
 from odoo import api, fields, models
-from odoo.exceptions import ValidationError
+from odoo.exceptions import ValidationError, UserError
 from odoo.fields import Command, Domain
-from odoo.tools import date_utils, float_compare, ormcache
+from odoo.tools import float_compare, ormcache
 from odoo.tools.date_utils import float_to_time, localized, to_timezone
 from odoo.tools.float_utils import float_round
 from odoo.tools.intervals import Intervals
 
-from odoo.addons.base.models.res_partner import _tz_get
-
-
-class DummyAttendance(NamedTuple):
-    hour_from: float
-    hour_to: float
-    dayofweek: str
-    day_period: str | None
-    week_type: str | None
-
 
 class ResourceCalendar(models.Model):
     """ Calendar model for a resource. It has
@@ -52,7 +41,6 @@ class ResourceCalendar(models.Model):
             company_id = res.get('company_id', self.env.company.id)
             company = self.env['res.company'].browse(company_id)
             res["attendance_ids"] = self._get_default_attendance_ids(company)
-            res["two_weeks_calendar"] = company.resource_calendar_id.two_weeks_calendar
         if 'full_time_required_hours' in fields and not res.get('full_time_required_hours'):
             company_id = res.get('company_id', self.env.company.id)
             company = self.env['res.company'].browse(company_id)
@@ -65,10 +53,6 @@ class ResourceCalendar(models.Model):
     attendance_ids = fields.One2many(
         'resource.calendar.attendance', 'calendar_id', 'Working Time',
         compute='_compute_attendance_ids', store=True, readonly=False, copy=True)
-    attendance_ids_1st_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 1st Week',
-        compute="_compute_two_weeks_attendance", inverse="_inverse_two_weeks_calendar")
-    attendance_ids_2nd_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 2nd Week',
-        compute="_compute_two_weeks_attendance", inverse="_inverse_two_weeks_calendar")
     company_id = fields.Many2one(
         'res.company', 'Company', domain=lambda self: [('id', 'in', self.env.companies.ids)],
         default=lambda self: self.env.company, index='btree_not_null')
@@ -76,22 +60,6 @@ class ResourceCalendar(models.Model):
     country_code = fields.Char(related='country_id.code', depends=['country_id'])
     leave_ids = fields.One2many(
         'resource.calendar.leaves', 'calendar_id', 'Time Off')
-    schedule_type = fields.Selection(
-        [
-            ('flexible', 'Flexible'),
-            ('fully_fixed', 'Fully Fixed'),
-        ],
-        string='Schedule Type',
-        required=True,
-        default='fully_fixed',
-        help="Choose which level of definition you want to define on your Schedule\n"
-            "- Flexible : Define an amount of hours to work on the week.\n"
-            "- Fully Fixed : define the days, periods and the start & end time for each period of the day",
-    )
-    duration_based = fields.Boolean("Attendance based on duration", help="The hours will be centered around 12:00 to cover the duration for the day")
-    flexible_hours = fields.Boolean(string="Flexible Hours",
-        compute="_compute_flexible_hours", inverse="_inverse_flexible_hours", store=True,
-        help="When enabled, it will allow employees to work flexibly, without relying on the company's working schedule (working hours).")
     full_time_required_hours = fields.Float(
         string="Full Time Equivalent",
         compute="_compute_full_time_required_hours", store=True, readonly=False,
@@ -107,13 +75,6 @@ class ResourceCalendar(models.Model):
         string="Hours per Week",
         compute="_compute_hours_per_week", store=True, readonly=False, copy=False)
     is_fulltime = fields.Boolean(compute='_compute_work_time_rate', string="Is Full Time")
-    two_weeks_calendar = fields.Boolean(string="Calendar in 2 weeks mode")
-    two_weeks_explanation = fields.Char('Explanation', compute="_compute_two_weeks_explanation")
-    tz = fields.Selection(
-        _tz_get, string='Timezone', required=True,
-        default=lambda self: self.env.context.get('tz') or self.env.user.tz or self.env.ref('base.user_admin').tz or 'UTC',
-        help="This field is used in order to define in which timezone the resources will work.")
-    tz_offset = fields.Char(compute='_compute_tz_offset', string='Timezone offset')
     work_resources_count = fields.Integer("Work Resources count", compute='_compute_work_resources_count')
     work_time_rate = fields.Float(string='Work Time Rate', compute='_compute_work_time_rate', search='_search_work_time_rate',
         help='Work time rate versus full time working schedule, should be between 0 and 100 %.')
@@ -125,82 +86,29 @@ class ResourceCalendar(models.Model):
     @api.constrains('attendance_ids')
     def _check_attendance_ids(self):
         for res_calendar in self:
-            if (res_calendar.two_weeks_calendar and
-                    res_calendar.attendance_ids.filtered(lambda a: a.display_type == 'line_section') and
-                    not res_calendar.attendance_ids.sorted('sequence')[0].display_type):
-                raise ValidationError(self.env._("In a calendar with 2 weeks mode, all periods need to be in the sections."))
-
             # Avoid superimpose in attendance
             attendance_ids = res_calendar.attendance_ids.filtered(
                 lambda attendance: not attendance.display_type)
-            if res_calendar.two_weeks_calendar:
-                res_calendar._check_overlap(attendance_ids.filtered(lambda attendance: attendance.week_type == '0'))
-                res_calendar._check_overlap(attendance_ids.filtered(lambda attendance: attendance.week_type == '1'))
-            else:
-                res_calendar._check_overlap(attendance_ids)
+            res_calendar._check_overlap(attendance_ids)
 
     # --------------------------------------------------
     # Compute Methods
     # --------------------------------------------------
 
-    @api.depends('two_weeks_calendar')
-    def _compute_two_weeks_attendance(self):
-        for calendar in self:
-            if not calendar.two_weeks_calendar:
-                continue
-            calendar.attendance_ids_1st_week = calendar.attendance_ids.filtered(lambda a: a.week_type == '0')
-            calendar.attendance_ids_2nd_week = calendar.attendance_ids.filtered(lambda a: a.week_type == '1')
-
-    def _inverse_two_weeks_calendar(self):
-        for calendar in self:
-            if not calendar.two_weeks_calendar:
-                continue
-            calendar.attendance_ids = calendar.attendance_ids_1st_week + calendar.attendance_ids_2nd_week
-
     @api.depends('hours_per_week', 'company_id.resource_calendar_id.hours_per_week')
     def _compute_full_time_required_hours(self):
         for calendar in self.filtered("company_id"):
             calendar.full_time_required_hours = calendar.company_id.resource_calendar_id.hours_per_week
 
-    @api.depends("schedule_type")
-    def _compute_flexible_hours(self):
-        for calendar in self:
-            calendar.flexible_hours = calendar.schedule_type == 'flexible'
-
-    def _inverse_flexible_hours(self):
-        for calendar in self:
-            calendar.schedule_type = 'flexible' if calendar.flexible_hours else 'fully_fixed'
-
     @api.depends('company_id')
     def _compute_attendance_ids(self):
         for calendar in self.filtered(lambda c: not c._origin or (c._origin.company_id != c.company_id and c.company_id)):
             company_calendar = calendar.company_id.resource_calendar_id
             calendar.update({
-                'two_weeks_calendar': company_calendar.two_weeks_calendar,
-                'tz': company_calendar.tz,
                 'attendance_ids': [(5, 0, 0)] + [
                     (0, 0, attendance._copy_attendance_vals()) for attendance in company_calendar.attendance_ids],
             })
 
-    @api.onchange('attendance_ids')
-    def _onchange_attendance_ids(self):
-        if not self.two_weeks_calendar:
-            return
-
-        even_week_seq = self.attendance_ids.filtered(lambda att: att.display_type == 'line_section' and att.week_type == '0')
-        odd_week_seq = self.attendance_ids.filtered(lambda att: att.display_type == 'line_section' and att.week_type == '1')
-        if len(even_week_seq) != 1 or len(odd_week_seq) != 1:
-            raise ValidationError(self.env._("You can't delete section between weeks."))
-
-        even_week_seq = even_week_seq.sequence
-        odd_week_seq = odd_week_seq.sequence
-
-        for line in self.attendance_ids.filtered(lambda att: att.display_type is False):
-            if even_week_seq > odd_week_seq:
-                line.week_type = '1' if even_week_seq > line.sequence else '0'
-            else:
-                line.week_type = '0' if odd_week_seq > line.sequence else '1'
-
     @api.depends('company_id')
     def _compute_global_leave_ids(self):
         for calendar in self.filtered(lambda c: not c._origin or c._origin.company_id != c.company_id):
@@ -209,37 +117,18 @@ class ResourceCalendar(models.Model):
                     (0, 0, leave._copy_leave_vals()) for leave in calendar.company_id.resource_calendar_id.global_leave_ids],
             })
 
-    @api.depends('attendance_ids', 'attendance_ids.hour_from', 'attendance_ids.hour_to', 'two_weeks_calendar', 'flexible_hours')
+    @api.depends('attendance_ids', 'attendance_ids.hour_from', 'attendance_ids.hour_to')
     def _compute_hours_per_day(self):
         """ Compute the average hours per day.
             Cannot directly depend on hours_per_week because of rounding issues. """
-        for calendar in self.filtered(lambda c: not c.flexible_hours):
+        for calendar in self:
             calendar.hours_per_day = float_round(calendar._get_hours_per_day(), precision_digits=2)
 
-    @api.depends('attendance_ids', 'attendance_ids.hour_from', 'attendance_ids.hour_to', 'two_weeks_calendar', 'flexible_hours')
+    @api.depends('attendance_ids', 'attendance_ids.hour_from', 'attendance_ids.hour_to')
     def _compute_hours_per_week(self):
         """ Compute the average hours per week """
-        for calendar in self.filtered(lambda c: not c.flexible_hours):
-            calendar.hours_per_week = float_round(calendar._get_hours_per_week(), precision_digits=2)
-
-    @api.depends('two_weeks_calendar')
-    def _compute_two_weeks_explanation(self):
-        today = fields.Date.today()
-        week_type = self.env['resource.calendar.attendance'].get_week_type(today)
-        week_type_str = self.env._("even") if week_type else self.env._("odd")
-        first_day = date_utils.start_of(today, 'week')
-        last_day = date_utils.end_of(today, 'week')
-        self.two_weeks_explanation = self.env._(
-            "The current week (from %(first_day)s to %(last_day)s) corresponds to %(number)s week.",
-            first_day=first_day,
-            last_day=last_day,
-            number=week_type_str,
-        )
-
-    @api.depends('tz')
-    def _compute_tz_offset(self):
         for calendar in self:
-            calendar.tz_offset = datetime.now(ZoneInfo(calendar.tz or 'GMT')).strftime('%z')
+            calendar.hours_per_week = float_round(calendar._get_hours_per_week(), precision_digits=2)
 
     def _compute_work_resources_count(self):
         resources_per_calendar = dict(self.env['resource.resource']._read_group(
@@ -289,77 +178,37 @@ class ResourceCalendar(models.Model):
         vals_list = super().copy_data(default=default)
         return [dict(vals, name=self.env._("%s (copy)", calendar.name)) for calendar, vals in zip(self, vals_list)]
 
-    # --------------------------------------------------
-    # Actions
-    # --------------------------------------------------
-
-    def switch_calendar_type(self):
-        self.ensure_one()
-        if not self.two_weeks_calendar:
-            self.two_weeks_calendar = True
-            final_attendances = self._get_two_weeks_attendance()
-            self.attendance_ids = [Command.clear()] + final_attendances
-
-        else:
-            self.two_weeks_calendar = False
-            self.attendance_ids.unlink()
-            self.duration_based = False
-            self.attendance_ids = self._get_default_attendance_ids(self.company_id)
-
-    def switch_based_on_duration(self):
-        self.ensure_one()
-        self.duration_based = not self.duration_based
-        if self.duration_based:
-            self.attendance_ids.filtered(lambda att: att.day_period == 'lunch').unlink()
-        else:
-            self.attendance_ids.unlink()
-            self.attendance_ids = self._get_default_attendance_ids(self.company_id)
-            if self.two_weeks_calendar:
-                self.attendance_ids = self._get_two_weeks_attendance()
-
     # --------------------------------------------------
     # Computation API
     # --------------------------------------------------
 
-    def _attendance_intervals_batch(self, start_dt, end_dt, resources=None, domain=None, tz=None, lunch=False):
+    def _attendance_intervals_batch(self, start_dt, end_dt, resources_per_tz=None, domain=None):
         assert start_dt.tzinfo and end_dt.tzinfo
-        self.ensure_one()
-        if not resources:
-            resources = self.env['resource.resource']
-            resources_list = [resources]
-        else:
-            resources_list = list(resources) + [self.env['resource.resource']]
+        if self:
+            self.ensure_one()
+        elif not resources_per_tz.values():
+            raise UserError(self.env._("You cannot compute the intervals without a calendar on no resource."))
 
-        if self.flexible_hours and lunch:
-            return {resource.id: Intervals([], keep_distinct=True) for resource in resources_list}
+        if not resources_per_tz:
+            resources_per_tz = {start_dt.tzinfo: self.env['resource.resource']}
 
         domain = Domain.AND([
             Domain(domain or Domain.TRUE),
             Domain('calendar_id', '=', self.id),
             Domain('display_type', '=', False),
-            Domain('day_period', '!=' if not lunch else '=', 'lunch'),
         ])
 
         attendances = self.env['resource.calendar.attendance'].search(domain)
-        # Since we only have one calendar to take in account
-        # Group resources per tz they will all have the same result
-        resources_per_tz = defaultdict(list)
-        for resource in resources_list:
-            resources_per_tz[tz or ZoneInfo((resource or self).tz)].append(resource)
+        duration_based_attendances = attendances.filtered('duration_based')
+        all_duration_based_attendances = attendances.calendar_id.attendance_ids.filtered('duration_based')
         # Resource specific attendances
         # Calendar attendances per day of the week
-        # * 7 days per week * 2 for two week calendars
-        attendances_per_day = [self.env['resource.calendar.attendance']] * 7 * 2
+        attendances_per_day = [self.env['resource.calendar.attendance']] * 7
         weekdays = set()
         for attendance in attendances:
             weekday = int(attendance.dayofweek)
             weekdays.add(weekday)
-            if self.two_weeks_calendar:
-                weektype = int(attendance.week_type)
-                attendances_per_day[weekday + 7 * weektype] |= attendance
-            else:
-                attendances_per_day[weekday] |= attendance
-                attendances_per_day[weekday + 7] |= attendance
+            attendances_per_day[weekday] |= attendance
 
         start = start_dt.astimezone(UTC)
         end = end_dt.astimezone(UTC)
@@ -373,15 +222,36 @@ class ResourceCalendar(models.Model):
             end = max(end, high.replace(tzinfo=UTC))
         # Generate once with utc as timezone
         days = rrule(DAILY, start.date(), until=end.date(), byweekday=weekdays)
-        ResourceCalendarAttendance = self.env['resource.calendar.attendance']
         base_result = []
         for day in days:
-            week_type = ResourceCalendarAttendance.get_week_type(day)
-            attendances = attendances_per_day[day.weekday() + 7 * week_type]
+            attendances = attendances_per_day[day.weekday()]
+
+            # If all attendance lines are duration based, compute correct intervals
+            if all(att.duration_based for att in attendances):
+                day_all_duration_based_attendances = all_duration_based_attendances.filtered(lambda att: int(att.dayofweek) == day.weekday())
+                day_duration_based_attendances = duration_based_attendances.filtered(lambda att: int(att.dayofweek) == day.weekday())
+                total_hours = sum(day_all_duration_based_attendances.mapped('duration_hours'))
+                hours_per_attendance = {}
+                current_hour_from = 12 - total_hours / 2
+                for att in day_all_duration_based_attendances:
+                    hours_per_attendance[att] = (current_hour_from, current_hour_from + att.duration_hours)
+                    current_hour_from += att.duration_hours
+                for att in day_duration_based_attendances:
+                    hour_from, hour_to = hours_per_attendance[att]
+                    day_from = datetime.combine(day, float_to_time(hour_from))
+                    day_to = datetime.combine(day, float_to_time(hour_to))
+                    base_result.append((day_from, day_to, attendance))
+                continue
+
             for attendance in attendances:
-                day_from = datetime.combine(day, float_to_time(attendance.hour_from))
-                day_to = datetime.combine(day, float_to_time(attendance.hour_to))
-                base_result.append((day_from, day_to, attendance))
+                if attendance.duration_based:
+                    day_from = datetime.combine(day, float_to_time(12 - attendance.duration_hours / 2))
+                    day_to = datetime.combine(day, float_to_time(12 + attendance.duration_hours / 2))
+                    base_result.append((day_from, day_to, attendance))
+                else:
+                    day_from = datetime.combine(day, float_to_time(attendance.hour_from))
+                    day_to = datetime.combine(day, float_to_time(attendance.hour_to))
+                    base_result.append((day_from, day_to, attendance))
 
         # Copy the result localized once per necessary timezone
         # Strictly speaking comparing start_dt < time or start_dt.astimezone(tz) < time
@@ -394,26 +264,30 @@ class ResourceCalendar(models.Model):
                     for val in base_result]
             for tz in resources_per_tz
         }
-        resource_calendars = resources._get_calendar_at(start_dt, tz)
         result_per_resource_id = dict()
         for tz, tz_resources in resources_per_tz.items():
             res = result_per_tz[tz]
+            calendar_data_by_resource = tz_resources._get_calendar_data_at(start_dt, tz)
 
             res_intervals = Intervals(res, keep_distinct=True)
             start_datetime = start_dt.astimezone(tz)
             end_datetime = end_dt.astimezone(tz)
 
-            for resource in tz_resources:
-                if resource and not resource_calendars.get(resource, False):
+            for resource in chain(tz_resources, (self.env['resource.resource'],) if self else ()):
+                calendar_data = calendar_data_by_resource.get(resource, {})
+                calendar = calendar_data.get('resource_calendar_id')
+                hours_per_week = calendar_data.get('hours_per_week')
+                hours_per_day = calendar_data.get('hours_per_day')
+                is_fully_flexible = not calendar and not hours_per_week and not hours_per_day
+                is_flexible = not calendar and (hours_per_week or hours_per_day)
+                if resource and is_fully_flexible:
                     # If the resource is fully flexible, return the whole period from start_dt to end_dt with a dummy attendance
                     hours = (end_dt - start_dt).total_seconds() / 3600
-                    days = hours / 24
                     dummy_attendance = self.env['resource.calendar.attendance'].new({
                         'duration_hours': hours,
-                        'duration_days': days,
                     })
                     result_per_resource_id[resource.id] = Intervals([(start_datetime, end_datetime, dummy_attendance)], keep_distinct=True)
-                elif self.flexible_hours or (resource and resource_calendars[resource].flexible_hours):
+                elif resource and is_flexible:
                     # For flexible Calendars, we create intervals to fill in the weekly intervals with the average daily hours
                     # until the full time required hours are met. This gives us the most correct approximation when looking at a daily
                     # and weekly range for time offs and overtime calculations and work entry generation
@@ -421,10 +295,8 @@ class ResourceCalendar(models.Model):
                     end_datetime_adjusted = end_datetime - relativedelta(seconds=1)
                     end_date = end_datetime_adjusted.date()
 
-                    calendar = resource_calendars[resource] if resource else self
-
-                    full_time_required_hours = calendar.full_time_required_hours
-                    max_hours_per_day = calendar.hours_per_day
+                    full_time_required_hours = hours_per_week
+                    max_hours_per_day = hours_per_day
 
                     intervals = []
                     current_start_day = start_date
@@ -468,7 +340,6 @@ class ResourceCalendar(models.Model):
 
                                 dummy_attendance = self.env['resource.calendar.attendance'].new({
                                     'duration_hours': allocate_hours,
-                                    'duration_days': 1,
                                 })
 
                                 intervals.append((start_time, end_time, dummy_attendance))
@@ -492,33 +363,35 @@ class ResourceCalendar(models.Model):
     def _leave_intervals(self, start_dt, end_dt, resource=None, domain=None, tz=None):
         if resource is None:
             resource = self.env['resource.resource']
-        return self._leave_intervals_batch(
-            start_dt, end_dt, resources=resource, domain=domain, tz=tz,
-        )[resource.id]
+        resources_per_tz = resource._get_resources_per_tz()
+        return self._leave_intervals_batch(start_dt, end_dt, resources_per_tz=resources_per_tz, domain=domain)[resource.id]
 
-    def _leave_intervals_batch(self, start_dt, end_dt, resources=None, domain=None, tz=None):
+    def _leave_intervals_batch(self, start_dt, end_dt, resources_per_tz=None, domain=None):
         """ Return the leave intervals in the given datetime range.
             The returned intervals are expressed in specified tz or in the calendar's timezone.
         """
         assert start_dt.tzinfo and end_dt.tzinfo
 
-        if not resources:
-            resources = self.env['resource.resource']
-            resources_list = [resources]
-        else:
-            resources_list = list(resources) + [self.env['resource.resource']]
         if domain is None:
             domain = [('time_type', '=', 'leave')]
         if self:
             domain = domain + [('calendar_id', 'in', [False] + self.ids)]
 
+        all_resources = set()
+        if not resources_per_tz or self:
+            all_resources.add(self.env['resource.resource'])
+            resources_per_tz = resources_per_tz or {start_dt.tzinfo: self.env['resource.resource']}
+        if resources_per_tz:
+            for _, resources in resources_per_tz.items():
+                all_resources |= set(resources)
+
         # for the computation, express all datetimes in UTC
         # Public leave don't have a resource_id
         domain = domain + [
-            ('resource_id', 'in', [False] + [r.id for r in resources_list]),
+            ('resource_id', 'in', [False] + [r.id for r in all_resources]),
             ('date_from', '<=', end_dt.astimezone(UTC).replace(tzinfo=None)),
             ('date_to', '>=', start_dt.astimezone(UTC).replace(tzinfo=None)),
-            ('company_id', 'in', [False] + ([r.company_id.id for r in resources_list if r.company_id] or [self.company_id.id])),
+            ('company_id', 'in', [False] + ([r.company_id.id for r in all_resources if r.company_id] or [self.company_id.id])),
         ]
 
         # retrieve leave intervals in (start_dt, end_dt)
@@ -530,74 +403,81 @@ class ResourceCalendar(models.Model):
             leave_company = leave.company_id
             leave_date_from = leave.date_from
             leave_date_to = leave.date_to
-            for resource in resources_list:
-                if leave_resource.id not in [False, resource.id] or (not leave_resource and resource and resource.company_id != leave_company):
-                    continue
-                tz = tz if tz else ZoneInfo((resource or self).tz)
-                if (tz, start_dt) in tz_dates:
-                    start = tz_dates[tz, start_dt]
-                else:
-                    start = start_dt.astimezone(tz)
-                    tz_dates[tz, start_dt] = start
-                if (tz, end_dt) in tz_dates:
-                    end = tz_dates[tz, end_dt]
-                else:
-                    end = end_dt.astimezone(tz)
-                    tz_dates[tz, end_dt] = end
-                dt0 = leave_date_from.astimezone(tz)
-                dt1 = leave_date_to.astimezone(tz)
-                if leave_resource and leave_resource._is_fully_flexible():
-                    dt0, dt1 = self._handle_flexible_leave_interval(dt0, dt1, leave)
-                result[resource.id].append((max(start, dt0), min(end, dt1), leave))
-
-        return {r.id: Intervals(result[r.id]) for r in resources_list}
-
-    def _work_intervals_batch(self, start_dt, end_dt, resources=None, domain=None, tz=None, compute_leaves=True):
+            for tz, resources in resources_per_tz.items():
+                for resource in chain(resources, (self.env['resource.resource'],) if self else ()):
+                    if leave_resource.id not in [False, resource.id] or (not leave_resource and resource and resource.company_id != leave_company):
+                        continue
+                    if (tz, start_dt) in tz_dates:
+                        start = tz_dates[tz, start_dt]
+                    else:
+                        start = start_dt.astimezone(tz)
+                        tz_dates[tz, start_dt] = start
+                    if (tz, end_dt) in tz_dates:
+                        end = tz_dates[tz, end_dt]
+                    else:
+                        end = end_dt.astimezone(tz)
+                        tz_dates[tz, end_dt] = end
+                    dt0 = leave_date_from.astimezone(tz)
+                    dt1 = leave_date_to.astimezone(tz)
+                    if leave_resource and leave_resource._is_fully_flexible():
+                        dt0, dt1 = self._handle_flexible_leave_interval(dt0, dt1, leave)
+                    result[resource.id].append((max(start, dt0), min(end, dt1), leave))
+
+        return {r.id: Intervals(result[r.id]) for r in all_resources}
+
+    def _work_intervals_batch(self, start_dt, end_dt, resources_per_tz=None, domain=None, compute_leaves=True):
         """ Return the effective work intervals between the given datetimes. """
-        if not resources:
-            resources = self.env['resource.resource']
-            resources_list = [resources]
-        else:
-            resources_list = list(resources) + [self.env['resource.resource']]
-
-        attendance_intervals = self._attendance_intervals_batch(start_dt, end_dt, resources, tz=tz or self.env.context.get("employee_timezone"))
+        all_resources = set()
+        if not resources_per_tz or self:
+            all_resources.add(self.env['resource.resource'])
+        if resources_per_tz:
+            for _, resources in resources_per_tz.items():
+                all_resources |= set(resources)
+
+        attendance_intervals = self._attendance_intervals_batch(start_dt, end_dt, resources_per_tz)
         if compute_leaves:
-            leave_intervals = self._leave_intervals_batch(start_dt, end_dt, resources, domain, tz=tz)
+            leave_intervals = self._leave_intervals_batch(start_dt, end_dt, resources_per_tz, domain)
             return {
-                r.id: (attendance_intervals[r.id] - leave_intervals[r.id]) for r in resources_list
+                r.id: (attendance_intervals[r.id] - leave_intervals[r.id]) for r in all_resources
             }
         return {
-            r.id: attendance_intervals[r.id] for r in resources_list
+            r.id: attendance_intervals[r.id] for r in all_resources
         }
 
     def _unavailable_intervals(self, start_dt, end_dt, resource=None, domain=None, tz=None):
         if resource is None:
             resource = self.env['resource.resource']
+        resources_per_tz = resource._get_resources_per_tz()
         return self._unavailable_intervals_batch(
-            start_dt, end_dt, resources=resource, domain=domain, tz=tz,
+            start_dt, end_dt, resources_per_tz=resources_per_tz, domain=domain,
         )[resource.id]
 
-    def _unavailable_intervals_batch(self, start_dt, end_dt, resources=None, domain=None, tz=None):
+    def _unavailable_intervals_batch(self, start_dt, end_dt, resources_per_tz=None, domain=None):
         """ Return the unavailable intervals between the given datetimes. """
-        if not resources:
-            resources = self.env['resource.resource']
-            resources_list = [resources]
-        else:
-            resources_list = list(resources)
-
-        resources_work_intervals = self._work_intervals_batch(start_dt, end_dt, resources, domain, tz)
+        all_resources = set()
+        if not resources_per_tz or self:
+            all_resources.add(self.env['resource.resource'])
+        if resources_per_tz:
+            for _, resources in resources_per_tz.items():
+                all_resources |= set(resources)
+        resources_work_intervals = self._work_intervals_batch(start_dt, end_dt, resources_per_tz, domain)
         result = {}
-        for resource in resources_list:
+        for resource in all_resources:
             if resource and resource._is_fully_flexible():
                 continue
-            work_intervals = [(start, stop) for start, stop, meta in resources_work_intervals[resource.id]]
-            # start + flatten(intervals) + end
-            work_intervals = [start_dt] + list(chain.from_iterable(work_intervals)) + [end_dt]
-            # put it back to UTC
-            work_intervals = [dt.astimezone(UTC) for dt in work_intervals]
-            # pick groups of two
-            work_intervals = list(zip(work_intervals[0::2], work_intervals[1::2]))
-            result[resource.id] = work_intervals
+            work_intervals = resources_work_intervals[resource.id]
+            utc_work_intervals = []
+            for (start, stop, meta) in work_intervals:
+                utc_work_intervals.append((start.astimezone(UTC), stop.astimezone(UTC), meta))
+
+            utc_work_intervals = Intervals(utc_work_intervals)
+            full_interval_UTC = Intervals([(
+                start_dt.astimezone(UTC),
+                end_dt.astimezone(UTC),
+                self.env['resource.calendar'],
+            )])
+
+            result[resource.id] = full_interval_UTC - utc_work_intervals
         return result
 
     # --------------------------------------------------
@@ -608,10 +488,18 @@ class ResourceCalendar(models.Model):
         """ attendance_ids correspond to attendance of a week,
             will check for each day of week that there are no superimpose. """
         result = []
+        hours_based_weekdays = set()
+        duration_based_weekdays = set()
         for attendance in attendance_ids:
-            # 0.000001 is added to each start hour to avoid to detect two contiguous intervals as superimposing.
-            # Indeed Intervals function will join 2 intervals with the start and stop hour corresponding.
-            result.append((int(attendance.dayofweek) * 24 + attendance.hour_from + 0.000001, int(attendance.dayofweek) * 24 + attendance.hour_to, attendance))
+            if attendance.duration_based and attendance.dayofweek in hours_based_weekdays or not attendance.duration_based and attendance.dayofweek in duration_based_weekdays:
+                raise ValidationError(self.env._("You cannot define hours and duration based attendances for the same day."))
+            if attendance.duration_based:
+                duration_based_weekdays.add(attendance.dayofweek)
+            else:
+                hours_based_weekdays.add(attendance.dayofweek)
+                # 0.000001 is added to each start hour to avoid to detect two contiguous intervals as superimposing.
+                # Indeed Intervals function will join 2 intervals with the start and stop hour corresponding.
+                result.append((int(attendance.dayofweek) * 24 + attendance.hour_from + 0.000001, int(attendance.dayofweek) * 24 + attendance.hour_to, attendance))
 
         if len(Intervals(result)) != len(result):
             raise ValidationError(self.env._("Attendances can't overlap."))
@@ -628,15 +516,18 @@ class ResourceCalendar(models.Model):
         """
         day_hours = defaultdict(float)
         day_days = defaultdict(float)
-        for start, stop, meta in attendance_intervals:
+        for start, stop, _ in attendance_intervals:
             # If the interval covers only a part of the original attendance, we
             # take durations in days proportionally to what is left of the interval.
             interval_hours = (stop - start).total_seconds() / 3600
             day_hours[start.date()] += interval_hours
-            if len(self) == 1 and self.flexible_hours:
-                day_days[start.date()] += interval_hours / self.hours_per_day if self.hours_per_day else 0
+
+        for day, hours in day_hours.items():
+            if len(self) == 1 and self._is_duration_based_on_date(day):
+                hours_per_day = self._get_duration_based_work_hours_on_date(day)
+                day_days[start.date()] += hours / hours_per_day if hours_per_day else 0
             else:
-                day_days[start.date()] += sum(meta.mapped('duration_days')) * interval_hours / sum(meta.mapped('duration_hours'))
+                day_days[day] = 0.5 if hours <= self.hours_per_day * 3 / 4 else 1
 
         return {
             # Round the number of days to the closest 16th of a day.
@@ -656,7 +547,7 @@ class ResourceCalendar(models.Model):
         def interval_dt(interval):
             return interval[1 if match_end else 0]
 
-        tz = resource.tz if resource else self.tz
+        tz = resource.tz if resource else 'UTC'
         if resource is None:
             resource = self.env['resource.resource']
 
@@ -673,8 +564,9 @@ class ResourceCalendar(models.Model):
 
         if not range_start <= dt <= range_end:
             return None
+        resources_per_tz = resource._get_resources_per_tz()
         work_intervals = sorted(
-            self._work_intervals_batch(range_start, range_end, resource, compute_leaves=compute_leaves)[resource.id],
+            self._work_intervals_batch(range_start, range_end, resources_per_tz, compute_leaves=compute_leaves)[resource.id],
             key=lambda i: abs(interval_dt(i) - dt),
         )
         return interval_dt(work_intervals[0]) if work_intervals else None
@@ -683,20 +575,18 @@ class ResourceCalendar(models.Model):
         # If the employee didn't work a full day, it is still counted, i.e. 19h / week (M/T/W(half day)) -> 3 days
         self.ensure_one()
         attendances = self._get_global_attendances()
-        if self.two_weeks_calendar:
-            number_of_days = len(set(attendances.filtered(lambda cal: cal.week_type == '1').mapped('dayofweek')))
-            number_of_days += len(set(attendances.filtered(lambda cal: cal.week_type == '0').mapped('dayofweek')))
-        else:
-            number_of_days = len(set(attendances.mapped('dayofweek')))
-        return number_of_days / 2 if self.two_weeks_calendar else number_of_days
+        return len(set(attendances.mapped('dayofweek')))
 
     def _get_hours_per_week(self):
         """ Calculate the average hours worked per week. """
         self.ensure_one()
         hour_count = 0.0
         for attendance in self._get_global_attendances():
-            hour_count += attendance.hour_to - attendance.hour_from
-        return hour_count / 2 if self.two_weeks_calendar else hour_count
+            if attendance.duration_based:
+                hour_count += attendance.duration_hours
+            else:
+                hour_count += attendance.hour_to - attendance.hour_from
+        return hour_count
 
     def _get_hours_per_day(self):
         """ Calculate the average hours worked per workday. """
@@ -705,14 +595,11 @@ class ResourceCalendar(models.Model):
         return hour_per_week / number_of_days if number_of_days else 0
 
     def _get_global_attendances(self):
-        return self.attendance_ids.filtered(lambda attendance:
-            attendance.day_period != 'lunch'
-            and not attendance.display_type)
+        return self.attendance_ids.filtered(lambda attendance: not attendance.display_type)
 
-    def _get_unusual_days(self, start_dt, end_dt, company_id=False):
-        if not self:
-            return {}
-        self.ensure_one()
+    def _get_unusual_days(self, start_dt, end_dt, company_id=False, resource=None):
+        if self:
+            self.ensure_one()
         if not start_dt.tzinfo:
             start_dt = start_dt.replace(tzinfo=UTC)
         if not end_dt.tzinfo:
@@ -721,8 +608,8 @@ class ResourceCalendar(models.Model):
         domain = []
         if company_id:
             domain = [('company_id', 'in', (company_id.id, False))]
-        if self.flexible_hours:
-            leave_intervals = self._leave_intervals_batch(start_dt, end_dt, domain=domain)[False]
+        if resource and resource._is_flexible():
+            leave_intervals = self._leave_intervals_batch(start_dt, end_dt, domain=domain, resources_per_tz=resource._get_resources_per_tz())[resource.id]
             works = set()
             for start_int, end_int, _ in leave_intervals:
                 works.update(start_int.date() + timedelta(days=i) for i in range((end_int.date() - start_int.date()).days + 1))
@@ -735,60 +622,22 @@ class ResourceCalendar(models.Model):
         if company_id and (attendances := company_id.resource_calendar_id.attendance_ids):
             return [
                 Command.create({
-                    'name': attendance.name,
                     'dayofweek': attendance.dayofweek,
-                    'week_type': attendance.week_type,
+                    'duration_hours': attendance.duration_hours,
                     'hour_from': attendance.hour_from,
                     'hour_to': attendance.hour_to,
-                    'day_period': attendance.day_period,
                     'display_type': attendance.display_type,
                 })
                 for attendance in attendances
             ]
         return [
-            Command.create({'name': self.env._('Monday Morning'), 'dayofweek': '0', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': self.env._('Monday Lunch'), 'dayofweek': '0', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': self.env._('Monday Afternoon'), 'dayofweek': '0', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
-            Command.create({'name': self.env._('Tuesday Morning'), 'dayofweek': '1', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': self.env._('Tuesday Lunch'), 'dayofweek': '1', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': self.env._('Tuesday Afternoon'), 'dayofweek': '1', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
-            Command.create({'name': self.env._('Wednesday Morning'), 'dayofweek': '2', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': self.env._('Wednesday Lunch'), 'dayofweek': '2', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': self.env._('Wednesday Afternoon'), 'dayofweek': '2', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
-            Command.create({'name': self.env._('Thursday Morning'), 'dayofweek': '3', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': self.env._('Thursday Lunch'), 'dayofweek': '3', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': self.env._('Thursday Afternoon'), 'dayofweek': '3', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
-            Command.create({'name': self.env._('Friday Morning'), 'dayofweek': '4', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': self.env._('Friday Lunch'), 'dayofweek': '4', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': self.env._('Friday Afternoon'), 'dayofweek': '4', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
+            Command.create({'dayofweek': '0', 'duration_hours': 8, 'hour_from': 0, 'hour_to': 0}),
+            Command.create({'dayofweek': '1', 'duration_hours': 8, 'hour_from': 0, 'hour_to': 0}),
+            Command.create({'dayofweek': '2', 'duration_hours': 8, 'hour_from': 0, 'hour_to': 0}),
+            Command.create({'dayofweek': '3', 'duration_hours': 8, 'hour_from': 0, 'hour_to': 0}),
+            Command.create({'dayofweek': '4', 'duration_hours': 8, 'hour_from': 0, 'hour_to': 0}),
         ]
 
-    def _get_two_weeks_attendance(self):
-        final_attendances = [
-            Command.create({
-                'name': 'First week',
-                'dayofweek': '0',
-                'sequence': '0',
-                'hour_from': 0,
-                'day_period': 'morning',
-                'week_type': '0',
-                'hour_to': 0,
-                'display_type':
-                'line_section'}),
-            Command.create({
-                'name': 'Second week',
-                'dayofweek': '0',
-                'sequence': '25',
-                'hour_from': 0,
-                'day_period': 'morning',
-                'week_type': '1',
-                'hour_to': 0,
-                'display_type': 'line_section'}),
-        ]
-        for idx, att in enumerate(self.attendance_ids):
-            final_attendances.append(Command.create(dict(att._copy_attendance_vals(), week_type='0', sequence=idx + 1)))
-            final_attendances.append(Command.create(dict(att._copy_attendance_vals(), week_type='1', sequence=idx + 26)))
-        return final_attendances
     # --------------------------------------------------
     # External API
     # --------------------------------------------------
@@ -804,11 +653,11 @@ class ResourceCalendar(models.Model):
             Counts the number of work hours between two datetimes.
         """
         self.ensure_one()
-        # Set timezone in UTC if no timezone is explicitly given
+        # Set timezone in company tz if no timezone is explicitly given
         if not start_dt.tzinfo:
-            start_dt = start_dt.replace(tzinfo=UTC)
+            start_dt = start_dt.astimezone(ZoneInfo(self.env.company.tz))
         if not end_dt.tzinfo:
-            end_dt = end_dt.replace(tzinfo=UTC)
+            end_dt = end_dt.astimezone(ZoneInfo(self.env.company.tz))
 
         if compute_leaves:
             intervals = self._work_intervals_batch(start_dt, end_dt, domain=domain)[False]
@@ -855,14 +704,15 @@ class ResourceCalendar(models.Model):
         Return datetime after having planned hours
         """
         revert = to_timezone(day_dt.tzinfo)
-        day_dt = localized(day_dt)
+        day_dt = day_dt.astimezone(ZoneInfo(resource.tz if resource else self.env.company.tz))
 
         if resource is None:
             resource = self.env['resource.resource']
 
         # which method to use for retrieving intervals
         if compute_leaves:
-            get_intervals = partial(self._work_intervals_batch, domain=domain, resources=resource)
+            resources_per_tz = resource._get_resources_per_tz()
+            get_intervals = partial(self._work_intervals_batch, domain=domain, resources_per_tz=resources_per_tz)
             resource_id = resource.id
         else:
             get_intervals = self._attendance_intervals_batch
@@ -900,7 +750,7 @@ class ResourceCalendar(models.Model):
         Returns the datetime of a days scheduling.
         """
         revert = to_timezone(day_dt.tzinfo)
-        day_dt = localized(day_dt)
+        day_dt = day_dt.astimezone(ZoneInfo(self.env.company.tz))
 
         # which method to use for retrieving intervals
         if compute_leaves:
@@ -938,72 +788,20 @@ class ResourceCalendar(models.Model):
 
         working_days = self._get_working_hours()
         dayofweek = str(date.weekday())
-        if self.two_weeks_calendar:
-            weektype = str(self.env['resource.calendar.attendance'].get_week_type(date))
-            return working_days[weektype][dayofweek]
-        return working_days[False][dayofweek]
+        return working_days[dayofweek]
 
-    def _get_hours_for_date(self, target_date, day_period=None):
-        """
-        An instance method on a calendar to get the start and end float hours for a given date.
-        :param target_date: The date to find working hours.
-        :param day_period: Optional string ('morning', 'afternoon') to filter for half-days.
-        :return: A tuple of floats (hour_from, hour_to).
-        """
+    def _is_duration_based_on_date(self, date):
         self.ensure_one()
-        if not target_date:
-            err = "Target Date cannot be empty"
-            raise ValueError(err)
-        if self.flexible_hours:
-            # Quick calculation to center flexible hours around 12PM midday
-            datetimes = [12.0 - self.hours_per_day / 2.0, 12.0, 12.0 + self.hours_per_day / 2.0]
-            if day_period:
-                return (datetimes[0], datetimes[1]) if day_period == 'morning' else (datetimes[1], datetimes[2])
-            return (datetimes[0], datetimes[2])
-
-        domain = [
-            ('calendar_id', '=', self.id),
-            ('display_type', '=', False),
-            ('day_period', '!=', 'lunch'),
-        ]
-
-        init_attendances = self.env['resource.calendar.attendance']._read_group(domain=domain,
-        groupby=['week_type', 'dayofweek', 'day_period'],
-        aggregates=['hour_from:min', 'hour_to:max'],
-        order='dayofweek,hour_from:min')
-
-        init_attendances = [DummyAttendance(hour_from, hour_to, dayofweek, day_period, week_type)
-            for week_type, dayofweek, day_period, hour_from, hour_to in init_attendances]
-
-        if day_period:
-            attendances = [att for att in init_attendances if att.day_period == day_period]
-            for attendance in filter(lambda att: att.day_period == 'full_day', init_attendances):
-                attendances.append(attendance._replace(
-                    hour_from=attendance.hour_from if day_period == 'morning' else 12,
-                    hour_to=attendance.hour_to if day_period == 'afternoon' else 12,
-                ))
-
-        else:
-            attendances = init_attendances
-
-        default_start = min((att.hour_from for att in attendances), default=0.0)
-        default_end = max((att.hour_to for att in attendances), default=0.0)
-
-        week_type = False
-        if self.two_weeks_calendar:
-            week_type = str(self.env['resource.calendar.attendance'].get_week_type(target_date))
-
-        filtered_attendances = [att for att in attendances if att.week_type == week_type and int(att.dayofweek) == target_date.weekday()]
-        hour_from = min((att.hour_from for att in filtered_attendances), default=default_start)
-        hour_to = max((att.hour_to for att in filtered_attendances), default=default_end)
+        return any(att.duration_based for att in self.attendance_ids.filtered(lambda a: int(a.dayofweek) == date.weekday()))
 
-        return (hour_from, hour_to)
+    def _get_duration_based_work_hours_on_date(self, date):
+        return sum(self.attendance_ids.filtered(lambda a: int(a.dayofweek) == date.weekday()).mapped('duration_hours'))
 
     @ormcache('self.id')
     def _get_working_hours(self):
         self.ensure_one()
 
-        working_days = defaultdict(lambda: defaultdict(lambda: False))
+        working_days = defaultdict(lambda: False)
         for attendance in self.attendance_ids:
-            working_days[attendance.week_type][attendance.dayofweek] = True
+            working_days[attendance.dayofweek] = True
         return working_days

--- a/addons/resource/models/resource_calendar_attendance.py
+++ b/addons/resource/models/resource_calendar_attendance.py
@@ -1,17 +1,12 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-import math
-
 from odoo import api, fields, models
-from odoo.exceptions import UserError
 
 
 class ResourceCalendarAttendance(models.Model):
     _name = 'resource.calendar.attendance'
     _description = "Work Detail"
-    _order = 'sequence, week_type, dayofweek, hour_from'
+    _order = 'sequence, dayofweek, hour_from'
 
-    name = fields.Char(required=True)
     dayofweek = fields.Selection([
         ('0', 'Monday'),
         ('1', 'Tuesday'),
@@ -27,106 +22,87 @@ class ResourceCalendarAttendance(models.Model):
     hour_to = fields.Float(string='Work to', default=0, required=True)
     # For the hour duration, the compute function is used to compute the value
     # unambiguously, while the duration in days is computed for the default
-    # value based on the day_period but can be manually overridden.
-    duration_hours = fields.Float(compute='_compute_duration_hours', inverse='_inverse_duration_hours', string='Duration (hours)', store=True, readonly=False)
-    duration_days = fields.Float(compute='_compute_duration_days', string='Duration (days)', store=True, readonly=False)
+    # value but can be manually overridden.
+    duration_hours = fields.Float(compute='_compute_duration_hours', string='Hours', store=True, readonly=False)
     calendar_id = fields.Many2one("resource.calendar", string="Resource's Calendar", required=True, index=True, ondelete='cascade')
-    duration_based = fields.Boolean(related="calendar_id.duration_based")
+    duration_based = fields.Boolean(compute='_compute_duration_based', store=True)
     day_period = fields.Selection([
         ('morning', 'Morning'),
-        ('lunch', 'Break'),
         ('afternoon', 'Afternoon'),
-        ('full_day', 'Full Day')], required=True, default='morning')
-    week_type = fields.Selection([
-        ('1', 'Second'),
-        ('0', 'First')
-        ], 'Week Number', default=False)
-    two_weeks_calendar = fields.Boolean("Calendar in 2 weeks mode", related='calendar_id.two_weeks_calendar')
+        ('full_day', 'Full Day')], store=True, compute='_compute_day_period')
     display_type = fields.Selection([
         ('line_section', "Section")], default=False, help="Technical field for UX purpose.")
     sequence = fields.Integer(default=10,
         help="Gives the sequence of this line when displaying the resource calendar.")
 
-    @api.onchange('hour_from', 'hour_to')
-    def _onchange_hours(self):
+    @api.onchange('hour_from')
+    def _onchange_hour_from(self):
         # avoid negative or after midnight
         self.hour_from = min(self.hour_from, 23.99)
         self.hour_from = max(self.hour_from, 0.0)
+
+    @api.onchange('hour_to')
+    def _onchange_hour_to(self):
+        # avoid negative or after midnight
         self.hour_to = min(self.hour_to, 24)
         self.hour_to = max(self.hour_to, 0.0)
 
+        if self.hour_from and not self.hour_to:
+            self.hour_from = 0.0
+
         # avoid wrong order
         self.hour_to = max(self.hour_to, self.hour_from)
 
-    @api.constrains('day_period')
-    def _check_day_period(self):
-        for attendance in self:
-            if attendance.day_period == 'lunch' and attendance.duration_based:
-                raise UserError(self.env._("%(att)s is a break attendance, You should not have such record on duration based calendar", att=attendance.name))
-
-    @api.model
-    def get_week_type(self, date):
-        # week_type is defined by
-        #  * counting the number of days from January 1 of year 1
-        #    (extrapolated to dates prior to the first adoption of the Gregorian calendar)
-        #  * converted to week numbers and then the parity of this number is asserted.
-        # It ensures that an even week number always follows an odd week number. With classical week number,
-        # some years have 53 weeks. Therefore, two consecutive odd week number follow each other (53 --> 1).
-        return int(math.floor((date.toordinal() - 1) / 7) % 2)
+    @api.onchange('duration_hours')
+    def _onchange_duration_hours(self):
+        self.duration_hours = min(self.duration_hours, 24)
+        if self.hour_from or self.hour_to:
+            if self.hour_from + self.duration_hours > 24:
+                self.hour_from = 24 - self.duration_hours
+                self.hour_to = 24
+            else:
+                self.hour_to = self.hour_from + self.duration_hours
 
     @api.depends('hour_from', 'hour_to')
-    def _compute_duration_hours(self):
-        for attendance in self.filtered('hour_to'):
-            attendance.duration_hours = (attendance.hour_to - attendance.hour_from) if attendance.day_period != 'lunch' else 0
-
-    def _inverse_duration_hours(self):
-        for calendar, attendances in self.grouped('calendar_id').items():
-            if not calendar.duration_based:
-                continue
-            for attendance in attendances:
-                if attendance.day_period == 'full_day':
-                    period_duration = attendance.duration_hours / 2
-                    attendance.hour_to = 12 + period_duration
-                    attendance.hour_from = 12 - period_duration
-                elif attendance.day_period == 'morning':
-                    attendance.hour_to = 12
-                    attendance.hour_from = 12 - attendance.duration_hours
-                elif attendance.day_period == 'afternoon':
-                    attendance.hour_to = 12 + attendance.duration_hours
-                    attendance.hour_from = 12
+    def _compute_duration_based(self):
+        for attendance in self:
+            attendance.duration_based = not attendance.hour_from and not attendance.hour_to
 
-    @api.depends('day_period')
-    def _compute_duration_days(self):
+    @api.depends('duration_hours', 'hour_from', 'hour_to')
+    def _compute_day_period(self):
         for attendance in self:
-            if attendance.day_period == 'lunch':
-                attendance.duration_days = 0
-            elif attendance.day_period == 'full_day':
-                attendance.duration_days = 1
+            if attendance.duration_hours > (0.75 * attendance.calendar_id.hours_per_day) or (not attendance.hour_from and not attendance.hour_to):
+                attendance.day_period = 'full_day'
+            elif attendance.hour_from and attendance.hour_to:
+                if attendance.hour_from > 12 or (12 - attendance.hour_from <= attendance.hour_to - 12):
+                    attendance.day_period = 'afternoon'
+                else:
+                    attendance.day_period = 'morning'
             else:
-                attendance.duration_days = 0.5 if attendance.duration_hours <= attendance.calendar_id.hours_per_day * 3 / 4 else 1
+                attendance.day_period = 'morning'
+
+    @api.depends('hour_from', 'hour_to')
+    def _compute_duration_hours(self):
+        for attendance in self.filtered(lambda att: att.hour_from or att.hour_to):
+            attendance.duration_hours = max(0, attendance.hour_to - attendance.hour_from)
 
-    @api.depends('week_type')
     def _compute_display_name(self):
         super()._compute_display_name()
-        this_week_type = str(self.get_week_type(fields.Date.context_today(self)))
-        section_names = {'0': self.env._('First week'), '1': self.env._('Second week')}
-        section_info = {True: self.env._('this week'), False: self.env._('other week')}
-        for record in self.filtered(lambda l: l.display_type == 'line_section'):
-            section_name = f"{section_names[record.week_type]} ({section_info[this_week_type == record.week_type]})"
-            record.display_name = section_name
+        dayofweek_selection = dict(self._fields['dayofweek']._description_selection(self.env))
+        day_period_selection = dict(self._fields['day_period']._description_selection(self.env))
+        for record in self.filtered(lambda l: not l.display_type):
+            record.display_name = f"{dayofweek_selection[record.dayofweek]} ({day_period_selection[record.day_period]})"
 
     def _copy_attendance_vals(self):
         self.ensure_one()
         return {
-            'name': self.name,
             'dayofweek': self.dayofweek,
+            'duration_hours': self.duration_hours,
             'hour_from': self.hour_from,
             'hour_to': self.hour_to,
-            'day_period': self.day_period,
-            'week_type': self.week_type,
             'display_type': self.display_type,
             'sequence': self.sequence,
         }
 
     def _is_work_period(self):
-        return self.day_period != 'lunch' and not self.display_type
+        return not self.display_type

--- a/addons/resource/models/resource_calendar_leaves.py
+++ b/addons/resource/models/resource_calendar_leaves.py
@@ -21,10 +21,7 @@ class ResourceCalendarLeaves(models.Model):
         if 'date_from' in fields and 'date_to' in fields and not res.get('date_from') and not res.get('date_to'):
             # Then we give the current day and we search the begin and end hours for this day in resource.calendar of the current company
             today = Datetime.now()
-            calendar = self.env.company.resource_calendar_id
-            if 'calendar_id' in res:
-                calendar = self.env['resource.calendar'].browse(res['calendar_id'])
-            tz = ZoneInfo(calendar.tz or 'UTC')
+            tz = ZoneInfo(self.env.company.tz or 'UTC')
             date_from = datetime.combine(today, time.min, tzinfo=tz)
             date_to = datetime.combine(today, time.max, tzinfo=tz)
             res.update(
@@ -65,7 +62,7 @@ class ResourceCalendarLeaves(models.Model):
     def _compute_date_to(self):
         user_tz = self.env.tz
         if not (self.env.user.tz or self.env.context.get('tz')):
-            user_tz = ZoneInfo(self.company_id.resource_calendar_id.tz or 'UTC')
+            user_tz = ZoneInfo(self.company_id.tz or 'UTC')
         for leave in self:
             if not leave.date_from or (leave.date_to and leave.date_to > leave.date_from):
                 continue

--- a/addons/resource/models/resource_mixin.py
+++ b/addons/resource/models/resource_mixin.py
@@ -22,6 +22,8 @@ class ResourceMixin(models.AbstractModel):
         'resource.calendar', 'Working Hours',
         default=lambda self: self.env.company.resource_calendar_id,
         index=True, related='resource_id.calendar_id', store=True, readonly=False)
+    hours_per_week = fields.Float(string="Hours per Week", related='resource_id.hours_per_week', readonly=False)
+    hours_per_day = fields.Float(string="Hours per Day", related='resource_id.hours_per_day', readonly=False)
     tz = fields.Selection(
         string='Timezone', related='resource_id.tz', readonly=False,
         help="This field is used in order to define in which timezone the resources will work.")
@@ -29,16 +31,19 @@ class ResourceMixin(models.AbstractModel):
     @api.model_create_multi
     def create(self, vals_list):
         resources_vals_list = []
-        calendar_ids = [vals['resource_calendar_id'] for vals in vals_list if vals.get('resource_calendar_id')]
-        calendars_tz = {calendar.id: calendar.tz for calendar in self.env['resource.calendar'].browse(calendar_ids)}
+        company_ids = [vals['company_id'] for vals in vals_list if vals.get('company_id')]
+        companies_tz = {company.id: company.tz for company in self.env['res.company'].browse(company_ids)}
         for vals in vals_list:
             if not vals.get('resource_id'):
                 resources_vals_list.append(
                     self._prepare_resource_values(
                         vals,
-                        vals.pop('tz', False) or calendars_tz.get(vals.get('resource_calendar_id'))
+                        vals.get('tz', False) or companies_tz.get(vals.get('company_id'))
                     )
                 )
+            elif not vals.get('tz'):
+                # Write the tz in the vals_list to forward to the version creation
+                vals['tz'] = self.env['resource.resource'].browse(vals.get('resource_id')).tz
         if resources_vals_list:
             resources = self.env['resource.resource'].create(resources_vals_list)
             resources_iter = iter(resources.ids)
@@ -80,6 +85,12 @@ class ResourceMixin(models.AbstractModel):
     def _get_calendars(self, date_from=None):
         return {resource.id: resource.resource_calendar_id for resource in self}
 
+    def _get_hours_per_week_batch(self, date_from=None):
+        return {resource.id: resource.hours_per_week for resource in self}
+
+    def _get_hours_per_day_batch(self, date_from=None):
+        return {resource.id: resource.hours_per_day for resource in self}
+
     def _get_work_days_data_batch(self, from_datetime, to_datetime, compute_leaves=True, calendar=None, domain=None):
         """
             By default the resource calendar is used, but it can be
@@ -108,16 +119,12 @@ class ResourceMixin(models.AbstractModel):
                 mapped_resources[calendar_by_resource[resource.id]] |= resource.resource_id
 
         for calendar, calendar_resources in mapped_resources.items():
-            if not calendar:
-                for calendar_resource in calendar_resources:
-                    result[calendar_resource.id] = {'days': 0, 'hours': 0}
-                continue
-
+            resources_per_tz = calendar_resources._get_resources_per_tz()
             # actual hours per day
             if compute_leaves:
-                intervals = calendar._work_intervals_batch(from_datetime, to_datetime, calendar_resources, domain)
+                intervals = calendar._work_intervals_batch(from_datetime, to_datetime, resources_per_tz, domain)
             else:
-                intervals = calendar._attendance_intervals_batch(from_datetime, to_datetime, calendar_resources)
+                intervals = calendar._attendance_intervals_batch(from_datetime, to_datetime, resources_per_tz)
 
             for calendar_resource in calendar_resources:
                 result[calendar_resource.id] = calendar._get_attendance_intervals_days_data(intervals[calendar_resource.id])
@@ -144,27 +151,28 @@ class ResourceMixin(models.AbstractModel):
         from_datetime = localized(from_datetime)
         to_datetime = localized(to_datetime)
 
-        mapped_resources = defaultdict(lambda: self.env['resource.resource'])
-        for record in self:
-            mapped_resources[calendar or record.resource_calendar_id] |= record.resource_id
+        resource_per_calendar = defaultdict(lambda: self.env['resource.resource'])
+        for resource in self:
+            resource_per_calendar[calendar or resource.resource_calendar_id] |= resource.resource_id
 
-        for calendar, calendar_resources in mapped_resources.items():
+        for calendar, resources in resource_per_calendar.items():
             # handle fully flexible resources by returning the length of the whole interval
             # since we do not take into account leaves for fully flexible resources
             if not calendar:
                 days = (to_datetime - from_datetime).days
                 hours = (to_datetime - from_datetime).total_seconds() / 3600
-                for calendar_resource in calendar_resources:
-                    result[calendar_resource.id] = {'days': days, 'hours': hours}
+                for resource in resources:
+                    result[resource.id] = {'days': days, 'hours': hours}
                 continue
 
             # compute actual hours per day
-            attendances = calendar._attendance_intervals_batch(from_datetime, to_datetime, calendar_resources)
-            leaves = calendar._leave_intervals_batch(from_datetime, to_datetime, calendar_resources, domain)
+            resources_per_tz = resources._get_resources_per_tz()
+            attendances = calendar._attendance_intervals_batch(from_datetime, to_datetime, resources_per_tz)
+            leaves = calendar._leave_intervals_batch(from_datetime, to_datetime, resources_per_tz, domain)
 
-            for calendar_resource in calendar_resources:
-                result[calendar_resource.id] = calendar._get_attendance_intervals_days_data(
-                    attendances[calendar_resource.id] & leaves[calendar_resource.id]
+            for resource in resources:
+                result[resource.id] = calendar._get_attendance_intervals_days_data(
+                    attendances[resource.id] & leaves[resource.id]
                 )
 
         # convert "resource: result" into "employee: result"
@@ -202,8 +210,8 @@ class ResourceMixin(models.AbstractModel):
         compute_leaves = self.env.context.get('compute_leaves', True)
 
         for calendar, records in records_by_calendar.items():
-            resources = self.resource_id
-            all_intervals = calendar._work_intervals_batch(from_datetime, to_datetime, resources, domain, compute_leaves=compute_leaves)
+            resources_per_tz = records.resource_id._get_resources_per_tz()
+            all_intervals = calendar._work_intervals_batch(from_datetime, to_datetime, resources_per_tz, domain, compute_leaves=compute_leaves)
             for record in records:
                 intervals = all_intervals[record.resource_id.id]
                 record_result = defaultdict(float)
@@ -211,31 +219,3 @@ class ResourceMixin(models.AbstractModel):
                     record_result[start.date()] += (stop - start).total_seconds() / 3600
                 result[record.id] = sorted(record_result.items())
         return result
-
-    def list_leaves(self, from_datetime, to_datetime, calendar=None, domain=None):
-        """
-            By default the resource calendar is used, but it can be
-            changed using the `calendar` argument.
-
-            `domain` is used in order to recognise the leaves to take,
-            None means default value ('time_type', '=', 'leave')
-
-            Returns a list of tuples (day, hours, resource.calendar.leaves)
-            for each leave in the calendar.
-        """
-        resource = self.resource_id
-        calendar = calendar or self.resource_calendar_id
-
-        # naive datetimes are made explicit in UTC
-        if not from_datetime.tzinfo:
-            from_datetime = from_datetime.replace(tzinfo=UTC)
-        if not to_datetime.tzinfo:
-            to_datetime = to_datetime.replace(tzinfo=UTC)
-
-        attendances = calendar._attendance_intervals_batch(from_datetime, to_datetime, resource)[resource.id]
-        leaves = calendar._leave_intervals_batch(from_datetime, to_datetime, resource, domain)[resource.id]
-        result = []
-        for start, stop, leave in (leaves & attendances):
-            hours = (stop - start).total_seconds() / 3600
-            result.append((start.date(), hours, leave))
-        return result

--- a/addons/resource/models/resource_resource.py
+++ b/addons/resource/models/resource_resource.py
@@ -9,6 +9,7 @@ from dateutil.relativedelta import relativedelta
 
 from odoo import api, fields, models
 from odoo.addons.base.models.res_partner import _tz_get
+from odoo.exceptions import ValidationError
 from odoo.tools import get_lang, babel_locale_parse
 from odoo.tools.intervals import Intervals
 from odoo.tools.date_utils import localized, sum_intervals, to_timezone, weeknumber, weekstart, weekend
@@ -50,10 +51,23 @@ class ResourceResource(models.Model):
         default=lambda self: self.env.company.resource_calendar_id,
         domain="[('company_id', '=', company_id)]",
         help="Define the working schedule of the resource. If not set, the resource will have fully flexible working hours.")
+    hours_per_week = fields.Float(string="Hours per Week", compute='_compute_hours_per_week', store=True, readonly=False)
+    hours_per_day = fields.Float(string="Hours per Day", compute='_compute_hours_per_day', store=True, readonly=False)
     tz = fields.Selection(
         _tz_get, string='Timezone', required=True,
         default=lambda self: self.env.context.get('tz') or self.env.user.tz or 'UTC')
 
+    @api.constrains('calendar_id', 'hours_per_week', 'hours_per_day')
+    def _check_hours_per_week_day(self):
+        for resource in self:
+            hours_per_week = resource.hours_per_week
+            hours_per_day = resource.hours_per_day
+            if not resource.calendar_id and hours_per_week and hours_per_day:
+                if hours_per_day < (hours_per_week / 7):
+                    raise ValidationError(self.env._("The average hours per day is too low compared to the number of hours per week"))
+                if hours_per_day > hours_per_week:
+                    raise ValidationError(self.env._("The average hours per day is too high compared to the number of hours per week"))
+
     _check_time_efficiency = models.Constraint(
         'CHECK(time_efficiency>0)',
         'Time efficiency must be strictly positive',
@@ -70,9 +84,8 @@ class ResourceResource(models.Model):
             if values.get('company_id') and not 'calendar_id' in values:
                 values['calendar_id'] = self.env['res.company'].browse(values['company_id']).resource_calendar_id.id
             if not values.get('tz'):
-                # retrieve timezone on user or calendar
-                tz = (self.env['res.users'].browse(values.get('user_id')).tz or
-                      self.env['resource.calendar'].browse(values.get('calendar_id')).tz)
+                # retrieve timezone on user
+                tz = self.env['res.users'].browse(values.get('user_id')).tz or self.env['res.company'].browse(values.get('company_id')).tz
                 if tz:
                     values['tz'] = tz
         return super().create(vals_list)
@@ -92,6 +105,17 @@ class ResourceResource(models.Model):
             return True
         return super().write(vals)
 
+    @api.depends('hours_per_week')
+    def _compute_hours_per_day(self):
+        for resource in self:
+            resource.hours_per_day = resource.hours_per_week / 7
+
+    @api.depends('hours_per_day')
+    def _compute_hours_per_week(self):
+        for resource in self:
+            if not resource.hours_per_week:
+                resource.hours_per_week = resource.hours_per_day * 7
+
     @api.onchange('company_id')
     def _onchange_company_id(self):
         if self.company_id:
@@ -158,7 +182,8 @@ class ResourceResource(models.Model):
         for calendar, resources in calendar_mapping.items():
             if not calendar:
                 continue
-            resources_unavailable_intervals = calendar._unavailable_intervals_batch(start_datetime, end_datetime, resources, tz=ZoneInfo(calendar.tz))
+            resources_per_tz = resources._get_resources_per_tz()
+            resources_unavailable_intervals = calendar._unavailable_intervals_batch(start_datetime, end_datetime, resources_per_tz)
             resource_mapping.update(resources_unavailable_intervals)
         return resource_mapping
 
@@ -208,7 +233,8 @@ class ResourceResource(models.Model):
                     resource_work_intervals[resource.id] |= Intervals([(start, end, self.env['resource.calendar.attendance'])])
                 continue
             # For each calendar used by the resources, retrieve the work intervals for every resources using it
-            work_intervals_batch = calendar._work_intervals_batch(start, end, resources=resources, compute_leaves=compute_leaves)
+            resources_per_tz = resources._get_resources_per_tz()
+            work_intervals_batch = calendar._work_intervals_batch(start, end, resources_per_tz=resources_per_tz, compute_leaves=compute_leaves)
             for resource in resources:
                 # Make the conjunction between work intervals and calendar validity
                 resource_work_intervals[resource.id] |= work_intervals_batch[resource.id] & resource_calendar_validity_intervals[resource.id][calendar]
@@ -219,17 +245,22 @@ class ResourceResource(models.Model):
     def _is_fully_flexible(self):
         """ employee has a fully flexible schedule has no working calendar set """
         self.ensure_one()
-        return not self.calendar_id
+        return bool(not self.calendar_id and not self.hours_per_week and not self.hours_per_day)
 
-    def _get_calendar_at(self, date_target, tz=False):
-        return {resource: resource.calendar_id for resource in self}
+    def _get_calendar_data_at(self, date_target, tz=False):
+        return {
+            resource: {
+                'resource_calendar_id': resource.calendar_id,
+                'hours_per_week': resource.hours_per_week,
+                'hours_per_day': resource.hours_per_day,
+            } for resource in self}
 
     def _is_flexible(self):
-        """ An employee is considered flexible if the field flexible_hours is True on the calendar
-            or the employee is not assigned any calendar, in which case is considered as Fully flexible.
+        """ An employee is considered flexible if he's not assigned any calendar
+            If the employee is not assigned any calendar and no hours_per_week or hours_per_day are defined, he's considered as Fully flexible.
         """
         self.ensure_one()
-        return self._is_fully_flexible() or (self.calendar_id and self.calendar_id.flexible_hours)
+        return self._is_fully_flexible() or bool(not self.calendar_id and self.hours_per_week and self.hours_per_day)
 
     def _get_flexible_resources_default_work_intervals(self, start, end):
         assert start.tzinfo and end.tzinfo
@@ -273,7 +304,7 @@ class ResourceResource(models.Model):
 
         while leave_start_day <= leave_end_day:
             if not self._is_fully_flexible():
-                hours = self.calendar_id.hours_per_day
+                hours = self.calendar_id.hours_per_day if self.calendar_id else self.hours_per_day
                 # only days inside the original period
                 if leave_start_day >= start_day and leave_start_day <= end_day:
                     resource_hours_per_day[self.id][leave_start_day] -= hours
@@ -329,8 +360,10 @@ class ResourceResource(models.Model):
                 # custom timeoff can divide a day to > 1 intervals
                 duration_per_day[day] += (interval_end - interval_start).total_seconds() / 3600
 
+            full_time_required_hours = resource.calendar_id.full_time_required_hours if resource.calendar_id else resource.company_id.resource_calendar_id.hours_per_week
+            hours_per_day = resource.calendar_id.hours_per_day if resource.calendar_id else resource.hours_per_day
             for day, hours in duration_per_day.items():
-                day_working_hours = min(hours, resource.calendar_id.hours_per_day)
+                day_working_hours = min(hours, hours_per_day)
                 # only days inside the original period
                 if day >= start_day and day <= end_day:
                     resource_hours_per_day[resource.id][day] = day_working_hours
@@ -338,11 +371,12 @@ class ResourceResource(models.Model):
                 year_week = weeknumber(babel_locale_parse(locale), day)
                 year, week = year_week
                 if (year < end_year) or (year == end_year and week <= end_week):
-                    resource_hours_per_week[resource.id][year_week] = min(resource.calendar_id.full_time_required_hours, day_working_hours + resource_hours_per_week[resource.id][year_week])
+                    resource_hours_per_week[resource.id][year_week] = min(full_time_required_hours, day_working_hours + resource_hours_per_week[resource.id][year_week])
 
         for calendar, resources in calendar_resources.items():
             domain = [('calendar_id', '=', False)] if not calendar else None
-            leave_intervals = (calendar or self.env['resource.calendar'])._leave_intervals_batch(min_start_date, max_end_date, resources, domain)
+            resources_per_tz = resources._get_resources_per_tz()
+            leave_intervals = (calendar or self.env['resource.calendar'])._leave_intervals_batch(min_start_date, max_end_date, resources_per_tz, domain)
             for resource_id, leaves in leave_intervals.items():
                 if not resource_id:
                     continue
@@ -396,3 +430,9 @@ class ResourceResource(models.Model):
                 work_hours_per_day[day] += day_working_hours
 
         return work_hours
+
+    def _get_resources_per_tz(self):
+        resources_per_tz = defaultdict(lambda: self.env['resource.resource'])
+        for resource in self:
+            resources_per_tz[ZoneInfo(resource.tz)] |= resource
+        return dict(resources_per_tz)
